lighthouse-core/audits/uses-rel-preconnect.js

```diff
@@ -0,0 +1,143 @@
+      description: 'Avoid multiple, costly round trips to any origin',
+      informative: true,
+      helpText:
+        'Consider using<link rel="preconnect dns-prefetch"> to set up early connections ' +
+        ' before an HTTP request is actually sent to the server. This will reduce multiple, ' +
+        `costly round trips to any origin. [Learn more](${learnMoreUrl}).`,
+      requiredArtifacts: ['devtoolsLogs'],
+      scoreDisplayMode: Audit.SCORING_MODES.NUMERIC,
+    };
+  }
+
+  /**
+   * Check is the connection is already open
+   * @param {!WebInspector.NetworkRequest} record
+   * @return {!boolean}
+   */
+  static hasAlreadyConnectedToOrigin(record) {
+    return (
+      record.timing &&
+      record.timing.dnsEnd - record.timing.dnsStart === 0 &&
+      record.timing.connectEnd - record.timing.connectStart === 0
+    );
+  }
+
+  /**
+   * Check is the connection has started before the socket idle time
+   * @param {!WebInspector.NetworkRequest} record
+   * @param {!WebInspector.NetworkRequest} mainResource
+   * @return {!boolean}
+   */
+  static socketStartTimeIsBelowThreshold(record, mainResource) {
+    return Math.max(0, record._startTime - mainResource._endTime) < PRECONNECT_SOCKET_MAX_IDLE;
+  }
+
+  /**
+   * @param {!Artifacts} artifacts
+   * @return {!AuditResult}
+   */
+  static async audit(artifacts) {
+    const devtoolsLogs = artifacts.devtoolsLogs[UsesRelPreconnectAudit.DEFAULT_PASS];
+    let maxWasted = 0;
+
+    const [networkRecords, mainResource] = await Promise.all([
+      artifacts.requestNetworkRecords(devtoolsLogs),
+      artifacts.requestMainResource(devtoolsLogs),
+    ]);
+
+    const origins = networkRecords
+      .filter(record => {
+        // filter out all resources that have the same origin
+        return (
+          !URL.originsMatch(mainResource.url, record.url) &&
+          // filter out all resources that are loaded by the document
+          record.initiatorRequest() !== mainResource &&
+          // filter out urls that do not have an origin (data, ...)
+          !!URL.getOrigin(record.url) &&
+          // filter out all resources where origins are already resolved
+          !UsesRelPreconnectAudit.hasAlreadyConnectedToOrigin(record) &&
+          // make sure the requests are below the PRECONNECT_SOCKET_MAX_IDLE (15s) mark
+          UsesRelPreconnectAudit.socketStartTimeIsBelowThreshold(record, mainResource)
+        );
+      })
+      .map(record => URL.getOrigin(record.url));
+
+    const preconnectOrigins = new Set(origins);
+    const results = [];
+    preconnectOrigins.forEach(origin => {
+      const records = networkRecords.filter(record => URL.getOrigin(record.url) === origin);
+      if (!records.length) {
+        return;
+      }
+
+      // Sometimes requests are done simultaneous and the connection has not been made
+      // chrome will try to connect for each network record, we get the first record
+      let firstRecordOfOrigin;
```

Comment:
@paulirish is this an acceptable occasion for reduce in your 👀 ? :)

this works too